### ISSUE TITLE
shell module in dh5view occasionally crashes when trying to display attributes for tab completion

### ISSUE THREAD

**Issue by David-Baddeley** [python-microscopy/python-microscopy#92]
**[Original report](https://bitbucket.org/david_baddeley/python-microscopy/issue/92) by Andrew Barentine (Bitbucket: [AndrewBarentine](https://bitbucket.org/AndrewBarentine), ).**

----------------------------------------

The resulting traceback is from running a recipe, trying to access the namespace by typing recipes.activeRecipe, but the following was thrown after typing “recipes.” with the period. I’ve seen this a couple of other times, but is usually solved by restarting dh5view and running the recipe again.

```
recipes.Traceback (most recent call last):
  File "C:\Users\Bergamot\Anaconda2\lib\site-packages\wx\py\shell.py", line 484, in OnChar
    self.autoCompleteShow(command)
  File "C:\Users\Bergamot\Anaconda2\lib\site-packages\wx\py\shell.py", line 1184, in autoCompleteShow
    includeDouble=self.autoCompleteIncludeDouble)
  File "C:\Users\Bergamot\Anaconda2\lib\site-packages\wx\py\interpreter.py", line 148, in getAutoCompleteList
    *args, **kwds)
  File "C:\Users\Bergamot\Anaconda2\lib\site-packages\wx\py\introspect.py", line 30, in getAutoCompleteList
    includeSingle, includeDouble)
  File "C:\Users\Bergamot\Anaconda2\lib\site-packages\pyface\util\fix_introspect_bug.py", line 80, in getAttributeNames
    attributes.sort(lambda x, y: cmp(x.upper(), y.upper()))
TypeError: must use keyword argument for key function
```

‌

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

What python, wx and pyface versions?

This looks like a python3 bug in pyface \(ie not our bug\) although it's vaguely familiar - I might have a work around somewhere. 

‌

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

What python, wx and pyface versions?

This looks like a python3 bug in pyface (ie not our bug) although it's
vaguely familiar - I might have a work around somewhere.

**Comment by David-Baddeley** [comment]
**Original comment by Zach Connerty-Marin (Bitbucket: [zacsimile](https://bitbucket.org/zacsimile), GitHub: [zacsimile](https://github.com/zacsimile)).**

----------------------------------------

And in VisGUI:  

```
pipeline.Traceback (most recent call last):
  File "C:\Users\zrc4\AppData\Local\conda\conda\envs\pyme3\lib\site-packages\wx\py\shell.py", line 484, in OnChar
    self.autoCompleteShow(command)
  File "C:\Users\zrc4\AppData\Local\conda\conda\envs\pyme3\lib\site-packages\wx\py\shell.py", line 1184, in autoCompleteShow
    includeDouble=self.autoCompleteIncludeDouble)
  File "C:\Users\zrc4\AppData\Local\conda\conda\envs\pyme3\lib\site-packages\wx\py\interpreter.py", line 148, in getAutoCompleteList
    *args, **kwds)
  File "C:\Users\zrc4\AppData\Local\conda\conda\envs\pyme3\lib\site-packages\wx\py\introspect.py", line 30, in getAutoCompleteList
    includeSingle, includeDouble)
  File "C:\Users\zrc4\AppData\Local\conda\conda\envs\pyme3\lib\site-packages\pyface\util\fix_introspect_bug.py", line 45, in getAttributeNames
    attributes += [attr for attr in special_attrs \
  File "C:\Users\zrc4\AppData\Local\conda\conda\envs\pyme3\lib\site-packages\pyface\util\fix_introspect_bug.py", line 46, in <listcomp>
    if hasattr(object, attr)]
  File "c:\users\zrc4\python-microscopy\PYME\LMVis\pipeline.py", line 328, in __getattr__
    return self.colourFilter[item]
  File "c:\users\zrc4\python-microscopy\PYME\IO\tabular.py", line 863, in __getitem__
    return self.get_channel_column(self.currentColour, keys)
  File "c:\users\zrc4\python-microscopy\PYME\IO\tabular.py", line 870, in get_channel_column
    return self.resultsSource[keys]
  File "c:\users\zrc4\python-microscopy\PYME\IO\tabular.py", line 544, in __getitem__
    return self.resultsSource[key][self.Index][sl]
  File "c:\users\zrc4\python-microscopy\PYME\IO\tabular.py", line 724, in __getitem__
    return self.resultsSource[keys]
  File "c:\users\zrc4\python-microscopy\PYME\IO\tabular.py", line 544, in __getitem__
    return self.resultsSource[key][self.Index][sl]
TypeError: only integer scalar arrays can be converted to a scalar index
r
```

wxpython 4.04

python 3.6.7

pyface 6.1.1

‌

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

I think Zachs error might be a bit different to Andrews. The biggest question is what pyface is doing there in the first place. We never import, wx doesn’t import it and `fix_introspect_bug` could better be termed `create_introspect_bug`. My best guess is the following - traits imports pyface for some unknown reason, and pyface takes it upon itself to monkey-patch `wx.py.shell`.

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

Andrew’s error has been fixed in the head of pyface \(we submitted a PR against it\) but as we rely on packages we’re some way off that. That bugfix won’t fix Zach’s error.

**Comment by David-Baddeley** [comment]
**Original comment by Zach Connerty-Marin (Bitbucket: [zacsimile](https://bitbucket.org/zacsimile), GitHub: [zacsimile](https://github.com/zacsimile)).**

----------------------------------------

I’m getting a similar error on Mac.

wxpython 4.04

python 3.6.8

pyface 6.0.0

‌

```
pipeline.Traceback (most recent call last):
  File "/anaconda3/envs/pyme/lib/python3.6/site-packages/wx/py/shell.py", line 484, in OnChar
    self.autoCompleteShow(command)
  File "/anaconda3/envs/pyme/lib/python3.6/site-packages/wx/py/shell.py", line 1184, in autoCompleteShow
    includeDouble=self.autoCompleteIncludeDouble)
  File "/anaconda3/envs/pyme/lib/python3.6/site-packages/wx/py/interpreter.py", line 148, in getAutoCompleteList
    *args, **kwds)
  File "/anaconda3/envs/pyme/lib/python3.6/site-packages/wx/py/introspect.py", line 30, in getAutoCompleteList
    includeSingle, includeDouble)
  File "/anaconda3/envs/pyme/lib/python3.6/site-packages/pyface/util/fix_introspect_bug.py", line 80, in getAttributeNames
    attributes.sort(lambda x, y: cmp(x.upper(), y.upper()))
TypeError: must use keyword argument for key function
```

‌

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

Error is now fixed in most recent pyface \(also in conda packages\), but has exposed another one. Have submitted a PR for the second error to the pyface project.

**Comment by barentine** [comment]
I haven't been running into any pyface issues in dh5view shell lately, but the visgui shell was still a problem with: traits 5.2, pyface-6.1.2-py36_0, traitsui-6.1.3-py_0, wxpython 4.0.4
```
>>> pipeline.recipe.Traceback (most recent call last):
  File "/home/smeagol/miniconda3/lib/python3.6/site-packages/wx/py/shell.py", line 484, in OnChar
    self.autoCompleteShow(command)
  File "/home/smeagol/miniconda3/lib/python3.6/site-packages/wx/py/shell.py", line 1184, in autoCompleteShow
    includeDouble=self.autoCompleteIncludeDouble)
  File "/home/smeagol/miniconda3/lib/python3.6/site-packages/wx/py/interpreter.py", line 148, in getAutoCompleteList
    *args, **kwds)
  File "/home/smeagol/miniconda3/lib/python3.6/site-packages/wx/py/introspect.py", line 30, in getAutoCompleteList
    includeSingle, includeDouble)
  File "/home/smeagol/miniconda3/lib/python3.6/site-packages/pyface/util/fix_introspect_bug.py", line 59, in getAttributeNames
    attrdict = getAllAttributeNames(object)
  File "/home/smeagol/miniconda3/lib/python3.6/site-packages/pyface/util/fix_introspect_bug.py", line 137, in getAllAttributeNames
    attrdict.update(getAllAttributeNames(klass))
  File "/home/smeagol/miniconda3/lib/python3.6/site-packages/pyface/util/fix_introspect_bug.py", line 137, in getAllAttributeNames
    attrdict.update(getAllAttributeNames(klass))
  File "/home/smeagol/miniconda3/lib/python3.6/site-packages/pyface/util/fix_introspect_bug.py", line 137, in getAllAttributeNames
    attrdict.update(getAllAttributeNames(klass))
  File "/home/smeagol/miniconda3/lib/python3.6/site-packages/pyface/util/fix_introspect_bug.py", line 144, in getAllAttributeNames
    if isinstance(bases, types.TupleType):
AttributeError: module 'types' has no attribute 'TupleType'
```

Building traits, traitsui, and pyface from source at their current heads however fixes it:
traits 6.1.0, traitsui 6.2.0, pyface 6.2.0

**Comment by zacsimile** [comment]
Fresh install of miniconda3 and pyme yields the following traceback.

```
ImportError: cannot import name 'Category'                                                                              

INFO:pyface.base_toolkit:No traitsui.toolkits plugin could be loaded for wx                                             

Traceback (most recent call last):                                                                                        

File "c:\users\zrc4\code\python-microscopy\PYME\LMVis\visCore.py", line 402, in OnOpenFile                                self.OpenFile(filename)                                                                                               

File "c:\users\zrc4\code\python-microscopy\PYME\LMVis\visCore.py", line 721, in OpenFile                                  self.recipeView._layout()                                                                                             

File "c:\users\zrc4\code\python-microscopy\PYME\recipes\vertical_recipe_display.py", line 122, in _layout                 pan = node.edit_traits(parent=item, kind=kind, view='pipeline_view_min')                                              

File "C:\Users\zrc4\AppData\Local\Continuum\miniconda3\envs\pyme\lib\site-packages\traits\has_traits.py", line 1706, in edit_traits                                                                                                               view = self.trait_view(view)                                                                                          

File "c:\users\zrc4\code\python-microscopy\PYME\recipes\base.py", line 209, in trait_view                                 import traitsui.api as tui                                                                                            

File "C:\Users\zrc4\AppData\Local\Continuum\miniconda3\envs\pyme\lib\site-packages\traitsui\api.py", line 36, in <module>                                                                                                                         from .editors.api import ArrayEditor                                                                                  

File "C:\Users\zrc4\AppData\Local\Continuum\miniconda3\envs\pyme\lib\site-packages\traitsui\editors\__init__.py", line 23, in <module>                                                                                                            from .api import ArrayEditor                                                                                          

File "C:\Users\zrc4\AppData\Local\Continuum\miniconda3\envs\pyme\lib\site-packages\traitsui\editors\api.py", line 24, in <module>                                                                                                                 from .code_editor import CodeEditor                                                                                   

File "C:\Users\zrc4\AppData\Local\Continuum\miniconda3\envs\pyme\lib\site-packages\traitsui\editors\code_editor.py", line 37, in <module>                                                                                                         class ToolkitEditorFactory(EditorFactory):                                                                            

File "C:\Users\zrc4\AppData\Local\Continuum\miniconda3\envs\pyme\lib\site-packages\traitsui\editors\code_editor.py", line 49, in ToolkitEditorFactory                                                                                             mark_color = Color(0xECE9D8)                                                                                          

File "C:\Users\zrc4\AppData\Local\Continuum\miniconda3\envs\pyme\lib\site-packages\traits\trait_factory.py", line 40, in __call__                                                                                                                 return self.maker_function(*args, **metadata)                                                                        

 File "C:\Users\zrc4\AppData\Local\Continuum\miniconda3\envs\pyme\lib\site-packages\traits\traits.py", line 717, in Color                                                                                                                          return ColorTrait(*args, **metadata)                                                                                  

File "C:\Users\zrc4\AppData\Local\Continuum\miniconda3\envs\pyme\lib\site-packages\traitsui\toolkit_traits.py", line 8, in ColorTrait                                                                                                             return toolkit().color_trait(*args, **traits)                                                                         

File "C:\Users\zrc4\AppData\Local\Continuum\miniconda3\envs\pyme\lib\site-packages\traitsui\toolkit.py", line 109, in toolkit                                                                                                                     _toolkit = find_toolkit('traitsui.toolkits')                                                                          

File "C:\Users\zrc4\AppData\Local\Continuum\miniconda3\envs\pyme\lib\site-packages\pyface\base_toolkit.py", line 263, in find_toolkit                                                                                                             return import_toolkit(ETSConfig.toolkit, entry_point)                                                                 

File "C:\Users\zrc4\AppData\Local\Continuum\miniconda3\envs\pyme\lib\site-packages\pyface\base_toolkit.py", line 227, in import_toolkit                                                                                                           raise RuntimeError(msg)                                                                                             

RuntimeError: No traitsui.toolkits plugin could be loaded for wx 
```
pyface 6.1.2
traitlets 4.3.3
traits 6.0.0
traitsui 6.1.3
wxpython 4.0.7.post2

**Comment by zacsimile** [comment]
Downgraded to

pyface 6.0.0
traits 5.1.1

I get:

```
Traceback (most recent call last):                                                                                       

File "C:\Users\zrc4\AppData\Local\Continuum\miniconda3\envs\pyme\lib\site-packages\traits\trait_notifiers.py", line 591, in _dispatch_change_event                                                                                              self.dispatch(handler, *args)                                                                                       

 File "C:\Users\zrc4\AppData\Local\Continuum\miniconda3\envs\pyme\lib\site-packages\traits\trait_notifiers.py", line 553, in dispatch                                                                                                            handler(*args)                                                                                                       

File "c:\users\zrc4\code\python-microscopy\PYME\ui\layerFoldPanel.py", line 95, in _refresh                              self.Refresh()                                                                                                     

RuntimeError: wrapped C/C++ object of type LayerCaptionBar has been deleted                                            ERROR:traits:Exception occurred in traits notification handler for object: <PYME.LMVis.layers.pointcloud.PointCloudRenderLayer object at 0x0000028570059468>, trait: trait_added, old value: <undefined>, new value: visble                   

Traceback (most recent call last):                                                                                       

File "C:\Users\zrc4\AppData\Local\Continuum\miniconda3\envs\pyme\lib\site-packages\traits\trait_notifiers.py", line 591, in _dispatch_change_event                                                                                              self.dispatch(handler, *args)                                                                                        

File "C:\Users\zrc4\AppData\Local\Continuum\miniconda3\envs\pyme\lib\site-packages\traits\trait_notifiers.py", line 553, in dispatch                                                                                                            handler(*args)                                                                                                       

File "c:\users\zrc4\code\python-microscopy\PYME\ui\layerFoldPanel.py", line 95, in _refresh                              self.Refresh()                                                                                                     

RuntimeError: wrapped C/C++ object of type LayerCaptionBar has been deleted                                            c:\users\zrc4\code\python-microscopy\PYME\ui\layerFoldPanel.py:46: wxPyDeprecationWarning: Call to deprecated item BitmapFromBuffer. Use :meth:`wx.Bitmap.FromBuffer` or :meth:`wx.Bitmap.FromBufferAndAlpha` instead.                          

return wx.BitmapFromBuffer(30, 10, img) 
```

**Comment by zacsimile** [comment]
Downgraded to wxpython=4.0.4 and it worked.

**Comment by David-Baddeley** [comment]
Was about to suggest downgrading wx. This is going to be a real pain in the proverbial if we need to pin to wx <=4.04 and correspondingly pin traitsui and/or pyface.

**Comment by David-Baddeley** [comment]
It's feeling like it might be time to fork traitsui (ugh!!!) and do all the wx4 fixes. I really don't wan't to though.

**Comment by zacsimile** [comment]
We could always switch to qt. Tossup if it'd be easier.

**Comment by zacsimile** [comment]
What is the status of this issue? Does the original error still occur? Do we just need to pin `wx` and `traitsui` versions? Do we replace this and and other relation issues with a superissue titled `Traits is a mess`?

**Comment by David-Baddeley** [comment]
current status is no good solutions :(. 

One potential workaround to this issue would be to force a reload of the module within wx that `fix_introspect_bug` patches before we use it, and hence remove the monkey-patching.

**Comment by zacsimile** [comment]
That sounds good. Is there a downside?

**Comment by barentine** [comment]
What's the issue with pinning to wx 4.0.4? You both know I hate being stuck in the past, but forking _more_ projects to migrate to py3(/wx4) given the expedient success with our own sounds like a nightmare. I'm not opposed to gui reworks as we've discussed (browser-based, etc.) but meanwhile: has anyone actually hit fix introspect / other issues with the following?
traits 5.1.2
traitsui 6.1.3
pyface 6.1.2
wxpython 4.0.4

**Comment by zacsimile** [comment]
It's worth noting that David's fix for this was released starting with `pyface==7.0.0` (https://docs.enthought.com/pyface/changelog.html), but `7.0.0` doesn't play nice with `wxpython==4.0.4`. Release `7.0.1` actually works with `wxpython==4.0.4` (you'll also need `traits==6.1.0` and `traitsui==7.0.1`). `7.0.0` and `7.0.1` are only available through `pip` at the moment (https://pypi.org/project/pyface/). I was able to get this installed and running on py3 using `requirements.txt` (plus installing `docutils` and `mpld3`). It's pure delight not having this introspection error. Thanks, @David-Baddeley!
